Libplanet/Blocks/BlockContentExtensions.cs

```diff
@@ -0,0 +1,55 @@
+        /// <summary>
+        /// Creates a <see cref="BlockContent{T}"/> by copying the fields of another block
+        /// <paramref name="content"/>.
+        /// <para>Note that it is a shortcut of <seealso cref="BlockContent{T}(IBlockContent{T})"/>
+        /// constructor.</para>
+        /// </summary>
+        /// <typeparam name="T">A class implementing <see cref="IAction"/> to include.  This type
+        /// parameter is aligned with <see cref="Transaction{T}"/>'s type parameter.</typeparam>
+        /// <param name="content">This source of the block content to copy.  This hasn't be
+        /// a actual <see cref="BlockContent{T}"/> instance, but can be any object which implements
+        /// <see cref="IBlockContent{T}"/> instance.</param>
+        /// <returns>The copied block content.</returns>
+        /// <exception cref="InvalidBlockProtocolVersionException">Thrown when
+        /// the <paramref name="content"/>'s to set is <see cref="IBlockMetadata.ProtocolVersion"/>
+        /// is less than 0, or greater than <see cref="BlockMetadata.CurrentProtocolVersion"/>,
+        /// the latest known protocol version.</exception>
+        /// <exception cref="InvalidBlockIndexException">Thrown when the value to set is negative.
+        /// </exception>
+        /// <exception cref="InvalidBlockDifficultyException">Thrown when the value to set is
+        ///  is negative.</exception>
+        /// <exception cref="InvalidBlockTotalDifficultyException">Thrown when
+        /// the <paramref name="content"/>'s <see cref="IBlockMetadata.TotalDifficulty"/> is less
+        /// than its <see cref="IBlockMetadata.Difficulty"/>.</exception>
+        /// <exception cref="InvalidTxSignatureException">Thrown when any tx signature is invalid or
+        /// not signed by its signer.</exception>
+        /// <exception cref="InvalidTxPublicKeyException">Thrown when any tx signer is not derived
+        /// from its its public key.</exception>
+        /// <exception cref="InvalidTxNonceException">Thrown when the same tx nonce is used by
+        /// a signer twice or more, or a tx nonce is used without its previous nonce by a signer.
+        /// Note that this validates only a block's intrinsic integrity between its transactions,
+        /// but does not guarantee integrity between blocks.  Such validation needs to be conducted
+        /// by <see cref="Blockchain.BlockChain{T}"/>.</exception>
+        /// <exception cref="InvalidTxGenesisHashException">Thrown when transactions to set have
+        /// inconsistent genesis hashes.</exception>
+        /// <exception cref="InvalidBlockTxHashException">Thrown when the given block
+        /// <paramref name="content"/>'s <see cref="IBlockMetadata.TxHash"/> is not consistent with
+        /// its <see cref="IBlockContent{T}.Transactions"/>.</exception>
+        /// <seealso cref="BlockContent{T}(IBlockContent{T})"/>
+        public static BlockContent<T> Copy<T>(this IBlockContent<T> content)
```

Comment:
Why is this needed? I can't seem to find its use. 🙁

Reply:
It's used in test code, and I think it's also useful for applications.

    <Libplanet.Tests>  (28 usages found)
        Blocks  (28 usages found)
            BlockContentExtensionsTest.cs  (3 usages found)
                (13: 55)  AssertBlockContentsEqual(Genesis, Genesis.Copy());
                (14: 53)  AssertBlockContentsEqual(Block1, Block1.Copy());
                (15: 57)  AssertBlockContentsEqual(BlockPv0, BlockPv0.Copy());
            BlockContentTest.cs  (3 usages found)
                (119: 32)  var block = Block1.Copy();
                (145: 32)  var block = Block1.Copy();
                (177: 53)  BlockContent<Arithmetic> block = Block1.Copy();
            BlockMetadataTest.cs  (1 usage found)
                (108: 38)  BlockMetadata b = Block1.Copy();
            PreEvaluationBlockHeaderTest.cs  (2 usages found)
                (187: 56)  BlockMetadata metadata = _contents.Genesis.Copy();
                (255: 56)  BlockMetadata metadata = _contents.Genesis.Copy();
            PreEvaluationBlockTest.cs  (19 usages found)
                (28: 66)  BlockContent<Arithmetic> content = _contents.Genesis.Copy();
                (36: 40)  content = _contents.Block1.Copy();
                (47: 40)  content = _contents.Block1.Copy();
                (53: 41)  content = _contents.Genesis.Copy();
                (59: 40)  content = _contents.Block1.Copy();
                (65: 41)  content = _contents.Genesis.Copy();
                (71: 41)  content = _contents.Genesis.Copy();
                (81: 66)  BlockContent<Arithmetic> content = _contents.Genesis.Copy();
                (93: 40)  content = _contents.Block1.Copy();
                (109: 40)  content = _contents.Block1.Copy();
                (131: 66)  BlockContent<Arithmetic> content = _contents.Genesis.Copy();
                (142: 40)  content = _contents.Block1.Copy();
                (165: 40)  content = _contents.Block1.Copy();
                (171: 41)  content = _contents.Genesis.Copy();
                (177: 40)  content = _contents.Block1.Copy();
                (183: 41)  content = _contents.Genesis.Copy();
                (189: 41)  content = _contents.Genesis.Copy();
                (202: 68)  BlockContent<Arithmetic> contentPv0 = _contents.Block1.Copy();
                (221: 68)  BlockContent<Arithmetic> contentPv1 = _contents.Block1.Copy();

Reply:
Yes, I had found it in the test code. I agree, it *may* be useful, but I think this can *also* be dangerous. I'm okay with copying `BlockMetadata<T>`, my anxiety might be unfounded, but I think being able to copy `BlockContent<T>` is a slightly different matter and may be able to cause a problem down the line. 🙁

Reply:
Is it a problem with this `.Copy()` extension method, or the `new BlockContent<T>(IBlockContent<T>)` constructor?

Reply:
Well, I think the question can be answered more easily with another question:

> Should we be able to copy `Block<T>` instances?

Ultimately, they are just reference copies anyway, it is hard to see any harm coming from it, and I'm probably just being paranoid, but who knows what might happen in the future? 😶

Reply:
`BlockContent<T>` in itself is not a block, and it does not have any proofs like nonce, pre-evaluation hash, or state-root hash.  It's similar to `BlockMetadata`, except it has `Transactions` too (besides `TxHash`).

Reply:
Unless I'm missing something, my thought process was `BlockMetadata` on it's own can't be consumed to create a valid `Block<T>` but `BlockContent<T>` can, hence the question above, and which is why if `Block<T>` shouldn't be copyable, neither should `PreEvaluationBlock<T>` and `BlockContent<T>`.

Reply:
If `Block<T>` shouldn't be copyable, it's because it's frozen with several proofs so that modifying few part of it is non-sense.  On the other hand, it's okay that `BlockContent<T>` & `BlockMetadata` can be copied (and changing few part of them), and it's also useful, because it's merely a draft of a block to produce.

In other perspective, copying immutable objects are not useful, because they are just duplicates.  However, copying mutable objects are useful, because in general, mutable objects change themselves in-place. In order to revise mutable objects with maintaining original contents at a time, they should be copyable.

Reply:
Ah, I get your point. Welp, I'm approaching from a more abstracted standpoint. Just being mutable doesn't mean copying and mutating won't have any side-effect to business-logic somewhere down the line. Anyway, I'm just talking fluffy nonsense. 😛